kapitan/refs/secrets/vaultkv.py

```diff
@@ -0,0 +1,285 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"hashicorp vault secrets module"
+
+import base64
+from binascii import Error as b_error
+import hashlib
+import re
+import os
+from sys import argv, exit
+
+import yaml
+
+import hvac
+from hvac.exceptions import Forbidden, InvalidPath
+from kapitan import cached
+from kapitan.errors import KapitanError
+from kapitan.resources import inventory_reclass
+from kapitan.refs.base64 import Base64Ref, Base64RefBackend
+from kapitan.refs.base import RefError
+
+try:
+    from yaml import CSafeLoader as YamlLoader
+except ImportError:
+    from yaml import SafeLoader as YamlLoader
+
+class VaultError(KapitanError):
+    """Generic vault errors"""
+    pass
+
+def get_env(parameter):
+    """
+    The following variables need to be exported to the environment depending on authentication needed
+    * VAULT_ADDR: url for vault
+    * VAULT_SKIP_VERIFY=true: if set, do not verify presented TLS certificate before communicating with Vault server.
+    * VAULT_TOKEN: token for vault
+    * VAULT_ROLE_ID: (required by approle)
+    * VAULT_SECRET_ID: (required by approle)
+    * VAULT_USERNAME: username to login to vault
+    * VAULT_PASSWORD: password to login to vault
+    * VAULT_CLIENT_KEY: path to an unencrypted PEM-encoded private key matching the client certificate
+    * VAULT_CLIENT_CERT: path to a PEM-encoded client certificate for TLS authentication to the Vault server
+    * VAULT_CACERT: path to a PEM-encoded CA cert file to use to verify the Vault server TLS certificate
+    * VAULT_CAPATH: path to a directory of PEM-encoded CA cert files to verify the Vault server TLS certificate
+    * VAULT_NAMESPACE: specify the Vault Namespace, if you have one
+    """
+    # Helper lambda function to get varibles either from parameters or environment
+    get_variable = lambda x: parameter.get(x, os.getenv('VAULT_'+x.upper(),default=''))
+    env = {}
+    env['url'] = get_variable('addr')
+    env['namespace'] = get_variable('namespace')
+    # VERIFY VAULT SERVER TLS CERTIFICATE
+    skip_verify = get_variable('skip_verify')
+    if skip_verify.lower() == 'false':
+        cert = get_variable('cacert')
+        if not cert:
+            cert_path = get_variable('capath')
+            if not cert_path:
+                raise Exception('Neither VAULT_CACERT nor VAULT_CAPATH specified')
+            env['verify'] = cert_path
+        else:
+            env['verify'] = cert
+
+    else:
+        env['verify'] = False
```

Comment:
Don't reuse a key for either a boolean or the actual cert or the cert path. 
The name should indicate exactly what's in that dictionary key. Otherwise other developers reviewing this code will have to re-run through what codepath was taken here and guess what sort of variable will be in this key.
This is how Type Confusion bugs occur. 
Quick random link found googling: https://nebelwelt.net/publications/files/18SyScan360-presentation.pdf

Reply:
`verify` is an `hvac.Client` parameter that expects either a boolean to indicate whether TLS verification should be performed when sending requests to Vault, or a string pointing at the CA bundle to use for verification.
I have added appropriate comments in the function.

Reply:
`verify` is an `hvac.Client` parameter that expects either a boolean to indicate whether TLS verification should be performed when sending requests to Vault, or a string pointing at the CA bundle to use for verification.
I have added appropriate comments in the function.